python/ray/tune/execution/tune_controller.py

```diff
@@ -519,6 +428,84 @@ def resume(self, resume_config: ResumeConfig):
 
             self.add_trial(trial_to_add)
 
+    def _restore_trials(self, experiment_state: Dict) -> List[Trial]:
+        trials = []
+        for trial_json_state, trial_runtime_metadata in experiment_state["trial_data"]:
+            trial = Trial.from_json_state(trial_json_state)
+            trial.restore_run_metadata(trial_runtime_metadata)
+
+            # The following properties may be updated on restoration
+            # Ex: moved local/cloud experiment directory
+
+            # Propagate updated storage ctx properties to the trial's restored copy.
+            new_storage = copy.copy(trial.storage)
+            new_storage.storage_filesystem = self._storage.storage_filesystem
+            new_storage.storage_fs_path = self._storage.storage_fs_path
+            new_storage.experiment_dir_name = self._storage.experiment_dir_name
+
+            # ATTN: `trial.set_storage` is used intentionally, since it
+            # also updates the absolute paths and filesystem of tracked checkpoints.
+            trial.set_storage(new_storage)
+
+            # Avoid creating logdir in client mode for returned trial results,
+            # since the dir might not be creatable locally.
+            # TODO(ekl) this is kind of a hack.
+            if not ray.util.client.ray.is_connected():
+                trial.init_local_path()  # Create logdir if it does not exist
```

Comment:
The naming is kind of convoluted now. In the future, we should have a unified naming, e.g. all use `staging` instead of `local`.